### ISSUE TITLE
NotifyPropertyChanged + Blazor Server UI Thread

### ISSUE THREAD

**Issue by garrettlondon1** [dotnet/aspnetcore#52217]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Blazor Server using Injectable view models fails to call StateHasChanged and influence the UI state when needed.

It's very unclear how to take advantage of ViewModel architecture in Blazor, when it comes to updating state of UI on main UI thread. It's easy to communicate between ViewModels with events, etc..

Example Repo:

https://github.com/garrettlondon1/MVVMExample

Issue:

- In ChildViewModel, Initialize is called, but IsLoading=true never triggers a state update. If you add **await Task.Delay(1)**, it will trigger a UI update
- If private setters can only be synchronous, how does Blazor Server even work with MVVM at all? The ViewModels will update, and **eventually** they will update UI state, but it doesn't happen when the developer wants it to happen.

### Expected Behavior

When Initializing ViewModel, setting IsLoading = true, **should** update UI state.

### Steps To Reproduce

In example repo:

- Run the application normally, you will see when re-initializing ChildViewModel on parent SelectedSection setter, the page does not reload.

- Comment out the await Task.Delay on the ChildViewModel initialize. The page does reload.
- It does not have to do with the amount of time that is being awaited. It has to do I believe with the fact that everything is synchronous in the viewmodel propertychanged notification.. and UI thread needs to wait on something. I have no idea, probably blazor internals.

### .NET Version

8.0.1

**Comment by MackinnonBuck** [comment]
Thanks for reaching out. We have an example in our E2E tests demonstrating how a component can utilize `INotifyPropertyChanged` [here](https://github.com/dotnet/aspnetcore/blob/836a61d9a4c89c0a8b27994c02f7f00dbe6cf1fa/src/Components/test/testassets/BasicTestApp/FormsTest/NotifyPropertyChangedValidationComponent.razor#L7). Let us know if this doesn't answer your question!

**Comment by garrettlondon1** [comment]
@MackinnonBuck I attached an example repository, I think the question is a little bit different..

The problem is that you cannot await a UI state change from a ViewModel, since the property notification is synchronous. The ViewModel cannot communicate to Blazor, in an async way, on the main UI thread, to re-render the component, even on property setter.

**Comment by garrettlondon1** [comment]
```
public Action        StateChanged      { get; set; }
public Func<Task>    StateChangedAsync { get; set; }
IViewModel SetUiManager(Action stateChanged, Func<Task> stateChangedAsync);
public IViewModel SetUiManager(Action stateChanged, Func<Task> stateChangedAsync)
        {
            _context.StateChanged = stateChanged;
            _context.StateChangedAsync = stateChangedAsync;
            return this;
        }
await VM.SetUiManager(StateHasChanged, () => InvokeAsync(StateHasChanged)).Initialize();
```

Here is one more way I have seen it solved, I think.. Instead of using Private Setters and NotifyPropertyChanged (edited) 

You can use a method before initialization, which passes async function into ViewModel, so you can callback and have more control over rendering
.. It eliminates all private setters
, but the problem is, now, you need methods which set properties and then call StateHasChanged
. So it really, only has benefit in the 1/100 times you cannot deal with rendering

**Comment by gragra33** [comment]
@garrettlondon1 This may help you: [Blazing.Mvvm - Blazor Extension for the MVVM CommunityToolkit](https://github.com/gragra33/Blazing.Mvvm/) and sample project: [Blazor MVVM Sample for CommunityToolkit.Mvvm using Blazing.Mvvm](https://github.com/gragra33/MvvmSampleBlazor)

**Comment by garrettlondon1** [comment]
> @garrettlondon1 This may help you: [Blazing.Mvvm - Blazor Extension for the MVVM CommunityToolkit](https://github.com/gragra33/Blazing.Mvvm/) and sample project: [Blazor MVVM Sample for CommunityToolkit.Mvvm using Blazing.Mvvm](https://github.com/gragra33/MvvmSampleBlazor)

It does help as a reference but the sample repo has an example that is pretty much using the same BaseViewModel as the package.

**Comment by gragra33** [comment]
@garrettlondon1 The library addresses all of your issues for you. The key thing the library does for you is to abstract the management of handing the `INotifyPropertyChanged` events and more for you. It is a wrapper for the MVVM Community Toolkit.

The samples demonstrate the simplicity of how to use. Each page covers the key features of the MVVM Community Toolkit: Observability, Commands, Messaging, Validation, etc... plus a demonstration of navigation by ViewModel with passing of data.

The second link with a sample project demonstrates moving a Xamarin mobile project using the MVVM Community Toolkit to a Blazor Wasm & Blazor Hybrid apps with minimal changes to the code. Something that is not easily done with other 3rd party Blazor MVVM libraries. Again demonstrating the simplicity of how the library works.

**Comment by garrettlondon1** [comment]
> @garrettlondon1 The library addresses all of your issues for you. The key thing the library does for you is to abstract the management of handing the `INotifyPropertyChanged` events and more for you. It is a wrapper for the MVVM Community Toolkit.
> 
> The samples demonstrate the simplicity of how to use. Each page covers the key features of the MVVM Community Toolkit: Observability, Commands, Messaging, Validation, etc... plus a demonstration of navigation by ViewModel with passing of data.
> 
> The second link with a sample project demonstrates moving a Xamarin mobile project using the MVVM Community Toolkit to a Blazor Wasm & Blazor Hybrid apps with minimal changes to the code. Something that is not easily done with other 3rd party Blazor MVVM libraries. Again demonstrating the simplicity of how the library works. So anyone who wants to move from Wpf/Xamarin/Maui to Blazor will be able to adapt quickly due to the similarity.
> 
> What sample were you expecting?
> 
> > The problem is that you cannot await a UI state change from a ViewModel, since the property notification is synchronous. The ViewModel cannot communicate to Blazor, in an async way, on the main UI thread, to re-render the component, even on property setter.
> 
> Here is how we handle it in the library:
> 
> 1. `MvvmComponentBase.cs`
> 
> ```cs
> public abstract class MvvmComponentBase<TViewModel> : ComponentBase, IView<TViewModel> where TViewModel : IViewModelBase
> {
>     [Inject]
>     protected TViewModel? ViewModel { get; set; }
> 
>     protected override void OnInitialized()
>     {
>         // Cause changes to the ViewModel to make Blazor re-render
>         ViewModel!.PropertyChanged += (_, _) => InvokeAsync(StateHasChanged);
>         base.OnInitialized();
>     }
> 
>     protected override Task OnInitializedAsync()
>         => ViewModel!.OnInitializedAsync();
> }
> ```
> 
> 2. `ViewModelBase.cs`
> 
> ```cs
> public abstract partial class ViewModelBase : ObservableObject, IViewModelBase
> {
>     public virtual async Task OnInitializedAsync()
>         => await Loaded().ConfigureAwait(true);
> 
>     protected virtual void NotifyStateChanged() => OnPropertyChanged((string?)null);
> 
>     [RelayCommand]
>     public virtual async Task Loaded()
>         => await Task.CompletedTask.ConfigureAwait(false);
> }
> ```
> 
> We give you full control through virtual methods of what code you implement asynchronously.

Correct, but have you tried with the ViewModel as a Scoped Service and not Transient? It seems like a Transient service plays nice with Blazor component disposal.

Scoped service does not, and when you clear items on a scoped service on Dispose().. The View will still be alive when the ViewModel is disposed.

Also, MVVM toolkit is way different than MVVM, I don't want to use the toolkit..

Basically,

View still alive while ViewModel is scoped, ViewModel clears properties with private setters. UI updates, View throws null reference exceptions, because all properties of ViewModel are cleared

**Comment by gragra33** [comment]
@garrettlondon1 MVVM Community Toolkit was built from MvvmLight with improvements. I'm not sure how it's not MVVM.

As for `Scoped` vs `Transient`, why `Scoped`? `Transient` is the traditional mode for Mvvm ViewModels. ViewModels are only meant to exist for the lifetime of the View. Don't try and push it beyond the design. Import `Scoped` state objects or services into the ViewModels if you require that type of data.

**Comment by garrettlondon1** [comment]
> @garrettlondon1 MVVM Community Toolkit was built from MvvmLight with improvements. I'm not sure how it's not MVVM.
> 
> As for `Scoped` vs `Transient`, why `Scoped`? `Transient` is the traditional mode for Mvvm ViewModels. ViewModels are only meant to exist for the lifetime of the View. Don't try and push it beyond the design. Inject `Scoped` state objects or services into the ViewModels if you require that type of data.

What do you do when you need to pass a child view model into the constructor of injected parent view model, but that view model is also created when the BaseView uses it.. Won't it instantiate twice? That means you cannot use dependency injection for child view models and subscribing to events?

**Comment by gragra33** [comment]
I would inject `ISeviceProvider` into the ParentViewModel and create a ChildViewModel using the injected `SeviceProvider` passing a reference to the ParentViewModel via a limited interface for the ChildViewModel to use. Again, the lifetime of the ChildViewModel should be `Transient`. When the ChildView (component) is destroyed, so should the ChildViewModel.

Another option is to use  `Messenger` to communicate between the two ViewModels. The sample code, that is part of the Blazing.Mvvm lib (first link above), shows you how to use `Messenger` to communicate between one or more ViewModels. Look at the `HexTranslate` page.

**Comment by garrettlondon1** [comment]
@gragra33 your help is much appreciated. It makes sense that ViewModels should be transient. I think the main curiosity from my side was how the synchronous nature of MVVM plays with the Blazor Server UI thread. But right now, it is getting more theoretical. Thanks very much.

**Comment by gragra33** [comment]
@garrettlondon1 I found some time over lunch to throw together an example for you: [Blazing.Mvvm.ParentChildSample](https://github.com/gragra33/Blazing.Mvvm.ParentChildSample)
